File: articles/what-the-meta-analyses-say-about-grit/julia/grit_meta_analysis_simulation.jl
# What the Meta-Analyses Say About Grit
# Synthetic meta-analysis workflow sketch.

using Random
using DataFrames
using CSV
using Statistics

Random.seed!(42)

k = 60
sample_size = rand(150:2500, k)

clampvec(x, lo, hi) = [min(max(v, lo), hi) for v in x]

r_total_grit = clampvec(0.18 .+ 0.08 .* randn(k), -0.10, 0.45)
r_perseverance = clampvec(0.22 .+ 0.08 .* randn(k), -0.10, 0.50)
r_consistency = clampvec(0.08 .+ 0.07 .* randn(k), -0.15, 0.35)

df = DataFrame(
    study_id = ["study_" * lpad(string(i), 2, "0") for i in 1:k],
    sample_size = sample_size,
    r_total_grit = r_total_grit,
    r_perseverance = r_perseverance,
    r_consistency = r_consistency
)

fisher_z(r) = 0.5 * log((1 + r) / (1 - r))
inverse_fisher_z(z) = (exp(2z) - 1) / (exp(2z) + 1)

function fixed_effect_meta_r(r_values, n_values)
    z_values = fisher_z.(r_values)
    variances = 1.0 ./ (n_values .- 3)
    weights = 1.0 ./ variances
    pooled_z = sum(weights .* z_values) / sum(weights)
    pooled_r = inverse_fisher_z(pooled_z)
    return pooled_r
end

summary = DataFrame(
    facet = ["total_grit", "perseverance_of_effort", "consistency_of_interests"],
    pooled_r = [
        fixed_effect_meta_r(r_total_grit, sample_size),
        fixed_effect_meta_r(r_perseverance, sample_size),
        fixed_effect_meta_r(r_consistency, sample_size)
    ]
)

mkpath(joinpath(@__DIR__, "..", "data", "processed"))
mkpath(joinpath(@__DIR__, "..", "outputs", "tables"))

CSV.write(joinpath(@__DIR__, "..", "data", "processed", "grit-meta-analysis-synthetic-studies-julia.csv"), df)
CSV.write(joinpath(@__DIR__, "..", "outputs", "tables", "grit-meta-analysis-summary-julia.csv"), summary)

println(summary)

File: articles/what-the-meta-analyses-say-about-grit/python/grit_meta_analysis_simulation.py
"""
What the Meta-Analyses Say About Grit

Synthetic study-level workflow for demonstrating pooled correlations,
Fisher's z transformation, and facet-level comparison.

This script is for article support and research-method demonstration only.
It is not a substitute for a real systematic review or meta-analysis.
"""

from pathlib import Path

import numpy as np
import pandas as pd


ROOT = Path(__file__).resolve().parents[1]
DATA_OUT = ROOT / "data" / "processed" / "grit-meta-analysis-synthetic-studies-python.csv"
SUMMARY_OUT = ROOT / "outputs" / "tables" / "grit-meta-analysis-pooled-effects-python.csv"

rng = np.random.default_rng(42)

k = 60
sample_size = rng.integers(150, 2500, size=k)

r_total_grit = np.clip(rng.normal(0.18, 0.08, size=k), -0.10, 0.45)
r_perseverance = np.clip(rng.normal(0.22, 0.08, size=k), -0.10, 0.50)
r_consistency = np.clip(rng.normal(0.08, 0.07, size=k), -0.15, 0.35)

outcome_domain = rng.choice(
    ["academic", "performance", "retention", "wellbeing"],
    size=k,
    replace=True,
    p=[0.45, 0.25, 0.20, 0.10],
)

df = pd.DataFrame(
    {
        "study_id": [f"study_{i+1:02d}" for i in range(k)],
        "sample_size": sample_size,
        "r_total_grit": r_total_grit,
        "r_perseverance": r_perseverance,
        "r_consistency": r_consistency,
        "outcome_domain": outcome_domain,
    }
)


def fisher_z(r: np.ndarray) -> np.ndarray:
    return 0.5 * np.log((1 + r) / (1 - r))


def inverse_fisher_z(z: float) -> float:
    return (np.exp(2 * z) - 1) / (np.exp(2 * z) + 1)


def fixed_effect_meta_r(r_values: np.ndarray, n_values: np.ndarray) -> dict:
    z_values = fisher_z(r_values)
    variances = 1 / (n_values - 3)
    weights = 1 / variances

    pooled_z = np.sum(weights * z_values) / np.sum(weights)
    pooled_se = np.sqrt(1 / np.sum(weights))

    lower_z = pooled_z - 1.96 * pooled_se
    upper_z = pooled_z + 1.96 * pooled_se

    return {
        "pooled_r": inverse_fisher_z(pooled_z),
        "ci_lower": inverse_fisher_z(lower_z),
        "ci_upper": inverse_fisher_z(upper_z),
    }


summary = pd.DataFrame(
    [
        {"facet": "total_grit", **fixed_effect_meta_r(df["r_total_grit"].to_numpy(), df["sample_size"].to_numpy())},
        {"facet": "perseverance_of_effort", **fixed_effect_meta_r(df["r_perseverance"].to_numpy(), df["sample_size"].to_numpy())},
        {"facet": "consistency_of_interests", **fixed_effect_meta_r(df["r_consistency"].to_numpy(), df["sample_size"].to_numpy())},
    ]
)

DATA_OUT.parent.mkdir(parents=True, exist_ok=True)
SUMMARY_OUT.parent.mkdir(parents=True, exist_ok=True)

df.to_csv(DATA_OUT, index=False)
summary.to_csv(SUMMARY_OUT, index=False)

print("Saved synthetic study data:", DATA_OUT)
print("Saved pooled effect summary:", SUMMARY_OUT)
print(summary.round(3))
